components/properties/src/provider/names.rs

```diff
@@ -39,48 +39,72 @@ use zerovec::{maps::ZeroMapKV, ule::VarULE, VarZeroSlice, VarZeroVec, ZeroMap, Z
 /// including in SemVer minor releases. While the serde representation of data structs is guaranteed
 /// to be stable, their Rust representation might not be. Use with caution.
 /// </div>
+///
+/// # Examples
+///
+/// Using a [`NormalizedPropertyNameStr`] as the key of a [`ZeroMap`]:
+///
+/// ```
+/// use icu_properties::provider::names::NormalizedPropertyNameStr;
+/// use zerovec::ZeroMap;
+///
+/// let map: ZeroMap<NormalizedPropertyNameStr, usize> = [
+///     (NormalizedPropertyNameStr::from_str("A_BC"), 11),
+///     (NormalizedPropertyNameStr::from_str("dEf"), 22),
+///     (NormalizedPropertyNameStr::from_str("G_H-I"), 33),
+/// ]
+/// .into_iter()
+/// .collect();
+///
+/// let key_approx = NormalizedPropertyNameStr::from_str("AB-C");
+/// let key_exact = NormalizedPropertyNameStr::from_str("A_BC");
+///
+/// // Strict lookup:
+/// assert_eq!(None, map.get_copied(key_approx));
+/// assert_eq!(Some(11), map.get_copied(key_exact));
+///
+/// // Loose lookup:
+/// assert_eq!(Some(11), map.get_copied_by(|u| u.cmp_loose(key_approx)));
+/// assert_eq!(Some(11), map.get_copied_by(|u| u.cmp_loose(key_exact)));
+/// ```
 #[derive(PartialEq, Eq)] // VarULE wants these to be byte equality
-#[derive(VarULE)]
+#[derive(Debug, VarULE)]
 #[repr(transparent)]
-pub struct NormalizedPropertyNameStr([u8]);
+pub struct NormalizedPropertyNameStr(UnvalidatedStr);
 
-#[cfg(feature = "datagen")]
+/// This impl requires enabling the optional `serde` Cargo feature of the `icu_properties` crate
+#[cfg(feature = "serde")]
 impl serde::Serialize for NormalizedPropertyNameStr {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: serde::Serializer,
     {
-        use serde::ser::Error;
-        if serializer.is_human_readable() {
-            let s = str::from_utf8(&self.0)
-                .map_err(|_| S::Error::custom("Attempted to datagen invalid string property"))?;
-            serializer.serialize_str(s)
-        } else {
-            serializer.serialize_bytes(&self.0)
-        }
+        self.0.serialize(serializer)
```

Comment:
question: does `derive` not work here (perhaps due to the DST?)

Reply:
`derive` works; thanks!